File: Problem 3/problem3.java
/**
 * Project Euler Problem 3
 * @author Tyler Smith
 * @email tsmith328@gatech.edu
 */

import java.util.List;
import java.util.ArrayList;

class problem3 {
    public static void main(String[] args) {
        List<Long> factors = getFactors(600851475143L);
        List<Long> primes = onlyPrimes(factors);
        long biggest = 0;
        for (long i : primes) {
            if (i > biggest) {
                biggest = i;
            }
        }
        System.out.println("The largest prime factor of that big number is:");
        System.out.println(biggest);
    }

    private static List<Long> getFactors(long num) {
        List<Long> factors = new ArrayList<Long>();
        for (long i = 1; i <= Math.sqrt(num); i++) {
            double k = ((double) num) / i;
            long kint = (long) k;
            if (k - kint < 0.00001 && k - kint > -0.0001) {
                factors.add(i);
                factors.add(kint);
            }
        }
        return factors;
    }

    private static List<Long> onlyPrimes(List<Long> factors){
        List<Long> primes = new ArrayList<Long>();
        for (long i : factors) {
            if (isPrime(i)) {
                primes.add(i);
            }
        }
        return primes;
    }

    private static boolean isPrime(long num) {
        if (num < 2) return false;
        if (num == 2) return true;
        for (long i = 1; i < (long) Math.sqrt(num) + 2; i++) {
            if (i != 1 && num % i == 0) {
                return false;
            }
        }
        return true;
    }
}

File: Problem 3/problem3.cpp
/*
 * Project Euler Problem 3
 * Tyler Smith
 * tsmith328@gatech.edu
 */

#include<iostream>
#include<list>
#include<cmath>

using std::cout;
using std::endl;
using std::list;

bool isPrime(long int);
void getFactors(long int, list<long int>*);
void onlyPrimes(const list<long int>*, list<long int>*);

int main()
{
    list<long int> factors;
    list<long int> primes;
    getFactors(600851475143L, &factors);
    onlyPrimes(&factors, &primes);
    list<long int>::const_iterator it;
    long int biggest = 0;
    for (it = primes.begin(); it != primes.end(); ++it)
    {
        if (*it > biggest)
        {
            biggest = *it;
        }
    }
    cout << "The largest prime factor of that big number is:" << endl;
    cout << biggest << endl;
}

void getFactors(long int num, list<long int>* output)
{
    for (long int i = 1; i <= sqrt((double) num); i++)
    {
        double k = ((double) num) / i;
        long int kint = (long int) k;
        if (k - kint < 0.0000000000000000001 && k - kint > -0.0000000000000000001)
        {
            output->push_back(i);
            output->push_back(kint);
        }
    }
}

void onlyPrimes(const list<long int>* input, list<long int>* output)
{
    list<long int>::const_iterator it;
    for (it = input->begin(); it!= input->end(); ++it)
    {
        if (isPrime(*it))
        {
            output->push_back(*it);
        }
    }
}

bool isPrime(long int num)
{
    if (num < 2) return false;
    if (num == 2) return true;
    for (long int i = 1; i < (long int) sqrt((double) num) + 2; i++)
    {
        if (i != 1 && num % i == 0)
        {
            return false;
        }
    }
    return true;
}